File: Sources/OPLFleetAgentCore/SessionScanner.swift
import Foundation

public actor SessionScanner {
  private static let timestampPrefix = Data("{\"timestamp\":\"".utf8)
  private static let newline = Data([0x0A])

  private struct FileCursor: Sendable {
    var offset: UInt64 = 0
    var remainder = Data()
    var parserState = TokenParserState()
    var lastModified = Date.distantPast
  }

  private struct SessionFile: Sendable {
    let url: URL
    let size: UInt64
    let modifiedAt: Date
  }

  public let codexHome: URL
  public let sessionsRoot: URL

  private let fileManager: FileManager
  private let calendar: Calendar
  private let activeFileHorizon: TimeInterval
  private let readChunkSize = 1_048_576
  private let markerOverlapSize = 4_096
  private var cursors: [URL: FileCursor] = [:]
  private var seenDeduplicationKeys: [String: Date] = [:]
  private var events: [UsageEvent] = []
  private var malformedRelevantLines = 0

  public init(
    codexHome: URL = SessionScanner.defaultCodexHome(),
    calendar: Calendar = .current,
    activeFileHorizon: TimeInterval = 65 * 60,
    fileManager: FileManager = .default
  ) {
    self.codexHome = codexHome
    sessionsRoot = codexHome.appendingPathComponent("sessions", isDirectory: true)
    self.calendar = calendar
    self.activeFileHorizon = activeFileHorizon
    self.fileManager = fileManager
  }

  public static func defaultCodexHome(
    environment: [String: String] = ProcessInfo.processInfo.environment
  ) -> URL {
    if let configured = environment["CODEX_HOME"]?.trimmingCharacters(in: .whitespacesAndNewlines),
      !configured.isEmpty
    {
      return URL(
        fileURLWithPath: NSString(string: configured).expandingTildeInPath, isDirectory: true)
    }
    return FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(
      ".codex", isDirectory: true)
  }

  public func refresh(now: Date = Date()) -> UsageSnapshot {
    var isDirectory: ObjCBool = false
    guard fileManager.fileExists(atPath: sessionsRoot.path, isDirectory: &isDirectory),
      isDirectory.boolValue
    else {
      return .empty(at: now, status: .sessionsDirectoryMissing)
    }

    do {
      let retentionStart = now.addingTimeInterval(-65 * 60)
      let files = try discoverSessionFiles(now: now)
      for file in files {
        try readAppendedContent(from: file, retentionStart: retentionStart)
      }

      events.removeAll { $0.timestamp < retentionStart }
      seenDeduplicationKeys = seenDeduplicationKeys.filter { $0.value >= retentionStart }
      let activeStart = now.addingTimeInterval(-2 * 60)
      let activeSessions = Set(files.filter { $0.modifiedAt >= activeStart }.map(\.url)).count

      return UsageMetricsCalculator.snapshot(
        events: events,
        now: now,
        activeSessions: activeSessions,
        malformedRelevantLines: malformedRelevantLines
      )
    } catch {
      return UsageSnapshot(
        generatedAt: now,
        oneMinute: .empty(windowSeconds: 60),
        fiveMinutes: .empty(windowSeconds: 300),
        thirtyMinutes: .empty(windowSeconds: 1_800),
        oneHour: .empty(windowSeconds: 3_600),
        activeSessions: 0,
        malformedRelevantLines: malformedRelevantLines,
        status: .readFailed
      )
    }
  }

  private func discoverSessionFiles(now: Date) throws -> [SessionFile] {
    let cutoff = now.addingTimeInterval(-activeFileHorizon)
    var files: [SessionFile] = []
    let resourceKeys: Set<URLResourceKey> = [
      .contentModificationDateKey, .fileSizeKey, .isRegularFileKey,
    ]
    guard let enumerator = fileManager.enumerator(
      at: sessionsRoot,
      includingPropertiesForKeys: Array(resourceKeys),
      options: [.skipsHiddenFiles, .skipsPackageDescendants]
    ) else {
      return files
    }

    for case let url as URL in enumerator where url.pathExtension == "jsonl" {
      let values = try url.resourceValues(forKeys: resourceKeys)
      guard values.isRegularFile == true else { continue }
      let modifiedAt = values.contentModificationDate ?? .distantPast
      guard modifiedAt >= cutoff || cursors[url] != nil else { continue }
      files.append(
        SessionFile(
          url: url,
          size: UInt64(max(values.fileSize ?? 0, 0)),
          modifiedAt: modifiedAt
        )
      )
    }

    return files.sorted { $0.url.path < $1.url.path }
  }

  private func readAppendedContent(from file: SessionFile, retentionStart: Date) throws {
    var cursor = cursors[file.url] ?? FileCursor()
    if file.size < cursor.offset {
      cursor = FileCursor()
    }

    guard file.size > cursor.offset else {
      cursor.lastModified = file.modifiedAt
      cursors[file.url] = cursor
      return
    }

    let handle = try FileHandle(forReadingFrom: file.url)
    defer { try? handle.close() }
    try handle.seek(toOffset: cursor.offset)
    let fallbackSessionID = file.url.deletingPathExtension().lastPathComponent
    let minimumTimestamp = retentionStart.addingTimeInterval(-2).formatted(.iso8601)

    while true {
      let appended = try autoreleasepool {
        try handle.read(upToCount: readChunkSize)
      }
      guard let appended, !appended.isEmpty else { break }

      autoreleasepool {
        var combined = cursor.remainder
        combined.append(appended)
        let split = splitRelevantCompleteLines(combined, minimumTimestamp: minimumTimestamp)
        cursor.remainder = split.remainder

        let batch = TokenEventParser.parse(
          lines: split.lines,
          state: &cursor.parserState,
          fallbackSessionID: fallbackSessionID
        )
        malformedRelevantLines += batch.malformedRelevantLines

        for event in batch.events where event.timestamp >= retentionStart {
          if seenDeduplicationKeys[event.deduplicationKey] == nil {
            seenDeduplicationKeys[event.deduplicationKey] = event.timestamp
            events.append(event)
          }
        }
      }
    }

    cursor.offset = file.size
    cursor.lastModified = file.modifiedAt
    cursors[file.url] = cursor
  }

  private func splitRelevantCompleteLines(
    _ data: Data,
    minimumTimestamp: String
  ) -> (lines: [Data], remainder: Data) {
    guard let finalNewline = data.lastIndex(of: 0x0A) else {
      return ([], retainedIncompleteLine(data))
    }

    let completeEnd = data.index(after: finalNewline)
    let completeRange = data.startIndex..<completeEnd
    var lineStarts: Set<Data.Index> = []

    for marker in TokenEventParser.relevantMarkers {
      var searchStart = completeRange.lowerBound
      while searchStart < completeRange.upperBound,
        let match = data.range(
          of: marker,
          options: [],
          in: searchStart..<completeRange.upperBound
        )
      {
        let lineStart =
          data.range(
            of: Self.newline,
            options: .backwards,
            in: data.startIndex..<match.lowerBound
          )?.upperBound ?? data.startIndex
        lineStarts.insert(lineStart)
        searchStart = match.upperBound
      }
    }

    let lines = lineStarts.sorted().compactMap { start -> Data? in
      guard
        let end = data.range(
          of: Self.newline,
          options: [],
          in: start..<completeEnd
        )?.lowerBound, start < end
      else {
        return nil
      }
      let line = Data(data[start..<end])
      return eventMayBeRecent(line, minimumTimestamp: minimumTimestamp) ? line : nil
    }
    let remainder =
      completeEnd < data.endIndex
      ? retainedIncompleteLine(data[completeEnd...])
      : Data()
    return (lines, remainder)
  }

  private func retainedIncompleteLine(_ data: Data) -> Data {
    if TokenEventParser.relevantMarkers.contains(where: { data.range(of: $0) != nil }) {
      return data
    }
    guard data.count > markerOverlapSize else { return data }
    return Data(data.suffix(markerOverlapSize))
  }

  private func eventMayBeRecent(_ line: Data, minimumTimestamp: String) -> Bool {
    guard line.starts(with: Self.timestampPrefix) else { return true }
    let start = line.index(line.startIndex, offsetBy: Self.timestampPrefix.count)
    guard let end = line[start...].firstIndex(of: 0x22) else { return true }
    let timestamp = String(decoding: line[start..<end], as: UTF8.self)
    return timestamp >= minimumTimestamp
  }
}

File: windows/src/OPLFleetAgent.Core/SessionScanner.cs
using System.Text;

namespace OPLFleetAgent.Core;

public sealed class SessionScanner
{
    private sealed class FileCursor
    {
        public long Offset { get; set; }
        public byte[] Remainder { get; set; } = [];
        public TokenParserState ParserState { get; } = new();
    }

    private sealed record SessionFile(string Path, long Size, DateTimeOffset ModifiedAt);

    private const int ReadChunkSize = 1_048_576;
    private const int MarkerOverlapSize = 4_096;
    private static readonly byte[][] MarkerBytes =
        TokenEventParser.RelevantMarkers.Select(Encoding.UTF8.GetBytes).ToArray();

    private readonly object gate = new();
    private readonly Dictionary<string, FileCursor> cursors =
        new(StringComparer.OrdinalIgnoreCase);
    private readonly Dictionary<string, DateTimeOffset> seenDeduplicationKeys =
        new(StringComparer.Ordinal);
    private readonly List<UsageEvent> events = [];
    private int malformedRelevantLines;

    public SessionScanner(string? codexHome = null)
    {
        CodexHome = ResolveCodexHome(codexHome);
        SessionsRoot = Path.Combine(CodexHome, "sessions");
    }

    public string CodexHome { get; }
    public string SessionsRoot { get; }

    public UsageSnapshot Refresh(DateTimeOffset? generatedAt = null)
    {
        lock (gate)
        {
            var now = generatedAt ?? DateTimeOffset.Now;
            if (!Directory.Exists(SessionsRoot))
            {
                return UsageSnapshot.Empty(now, CollectionStatus.SessionsDirectoryMissing);
            }

            try
            {
                var retentionStart = now.AddMinutes(-65);
                var files = DiscoverSessionFiles(now);
                foreach (var file in files)
                {
                    ReadAppendedContent(file, retentionStart);
                }

                events.RemoveAll(item => item.Timestamp < retentionStart);
                foreach (var key in seenDeduplicationKeys
                    .Where(item => item.Value < retentionStart)
                    .Select(item => item.Key)
                    .ToArray())
                {
                    seenDeduplicationKeys.Remove(key);
                }

                var activeStart = now.AddMinutes(-2);
                var activeSessions = files.Count(file => file.ModifiedAt >= activeStart);
                return UsageMetrics.Snapshot(
                    events,
                    now,
                    activeSessions,
                    malformedRelevantLines);
            }
            catch (IOException)
            {
                return UsageSnapshot.Empty(now, CollectionStatus.ReadFailed) with
                {
                    MalformedRelevantLines = malformedRelevantLines,
                };
            }
            catch (UnauthorizedAccessException)
            {
                return UsageSnapshot.Empty(now, CollectionStatus.ReadFailed) with
                {
                    MalformedRelevantLines = malformedRelevantLines,
                };
            }
        }
    }

    public static string ResolveCodexHome(string? explicitPath = null)
    {
        var configured = string.IsNullOrWhiteSpace(explicitPath)
            ? Environment.GetEnvironmentVariable("CODEX_HOME")
            : explicitPath;
        if (string.IsNullOrWhiteSpace(configured))
        {
            return Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                ".codex");
        }

        var expanded = Environment.ExpandEnvironmentVariables(configured.Trim().Trim('"'));
        if (expanded == "~")
        {
            expanded = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        }
        else if (expanded.StartsWith($"~{Path.DirectorySeparatorChar}", StringComparison.Ordinal))
        {
            expanded = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                expanded[2..]);
        }
        return Path.GetFullPath(expanded);
    }

    private List<SessionFile> DiscoverSessionFiles(DateTimeOffset now)
    {
        var cutoff = now.AddMinutes(-65);
        var files = new List<SessionFile>();
        foreach (var path in Directory.EnumerateFiles(
                     SessionsRoot,
                     "*.jsonl",
                     SearchOption.AllDirectories))
        {
            var info = new FileInfo(path);
            var modifiedAt = new DateTimeOffset(info.LastWriteTimeUtc, TimeSpan.Zero);
            if (modifiedAt < cutoff && !cursors.ContainsKey(path))
            {
                continue;
            }
            files.Add(new SessionFile(path, Math.Max(info.Length, 0), modifiedAt));
        }
        return files.OrderBy(item => item.Path, StringComparer.OrdinalIgnoreCase).ToList();
    }

    private void ReadAppendedContent(SessionFile file, DateTimeOffset retentionStart)
    {
        if (!cursors.TryGetValue(file.Path, out var cursor) || file.Size < cursor.Offset)
        {
            cursor = new FileCursor();
            cursors[file.Path] = cursor;
        }
        if (file.Size <= cursor.Offset)
        {
            return;
        }

        using var stream = new FileStream(
            file.Path,
            FileMode.Open,
            FileAccess.Read,
            FileShare.ReadWrite | FileShare.Delete,
            ReadChunkSize,
            FileOptions.SequentialScan);
        stream.Seek(cursor.Offset, SeekOrigin.Begin);
        var buffer = new byte[ReadChunkSize];
        while (true)
        {
            var count = stream.Read(buffer, 0, buffer.Length);
            if (count <= 0)
            {
                break;
            }

            var combined = new byte[cursor.Remainder.Length + count];
            cursor.Remainder.CopyTo(combined, 0);
            buffer.AsSpan(0, count).CopyTo(combined.AsSpan(cursor.Remainder.Length));
            var split = SplitRelevantCompleteLines(combined);
            cursor.Remainder = split.Remainder;
            var batch = TokenEventParser.Parse(
                split.Lines,
                cursor.ParserState,
                Path.GetFileNameWithoutExtension(file.Path));
            malformedRelevantLines += batch.MalformedRelevantLines;
            foreach (var usageEvent in batch.Events.Where(item => item.Timestamp >= retentionStart))
            {
                if (seenDeduplicationKeys.TryAdd(
                    usageEvent.DeduplicationKey,
                    usageEvent.Timestamp))
                {
                    events.Add(usageEvent);
                }
            }
        }
        cursor.Offset = stream.Position;
    }

    private static (IReadOnlyList<string> Lines, byte[] Remainder) SplitRelevantCompleteLines(
        byte[] data)
    {
        var finalNewline = Array.LastIndexOf(data, (byte)'\n');
        if (finalNewline < 0)
        {
            return ([], RetainIncompleteLine(data));
        }

        var lines = new List<string>();
        var lineStart = 0;
        while (lineStart <= finalNewline)
        {
            var lineEnd = Array.IndexOf(data, (byte)'\n', lineStart, finalNewline - lineStart + 1);
            if (lineEnd < 0)
            {
                break;
            }
            var line = data.AsSpan(lineStart, lineEnd - lineStart);
            if (line.Length > 0 && ContainsRelevantMarker(line))
            {
                lines.Add(Encoding.UTF8.GetString(line));
            }
            lineStart = lineEnd + 1;
        }

        return (lines, data[(finalNewline + 1)..]);
    }

    private static byte[] RetainIncompleteLine(byte[] data)
    {
        if (ContainsRelevantMarker(data) || data.Length <= MarkerOverlapSize)
        {
            return data;
        }
        return data[^MarkerOverlapSize..];
    }

    private static bool ContainsRelevantMarker(ReadOnlySpan<byte> line)
    {
        foreach (var marker in MarkerBytes)
        {
            if (line.IndexOf(marker) >= 0)
            {
                return true;
            }
        }
        return false;
    }
}
